### ISSUE TITLE
Horizontal Legends

### ISSUE THREAD

**Issue by johanndt** [sandialabs/toyplot#161]
Hi again,

I've been asked by a client to create some legends at the top which are displayed horizontally across the top. Something like this:

![image](https://cloud.githubusercontent.com/assets/7507499/16890324/a8a79d7c-4b30-11e6-9207-6e76298a4564.png)

I've looked through the code and it appears you can only really put the legends labels vertically one after the other. I was looking around here: https://github.com/sandialabs/toyplot/blob/d4b22d9b297fc35b6605f805daed08a528f774fd/toyplot/html.py#L2123

Now, before I go off and code something, I just wanted to make sure there isn't some fancy styling or something else which can achieve the same result?

Otherwise I'm happy to have a go at it. What would be the best approach? Adding a new parameter and trickling that through to the render?

Or my personal favourite might be to just check that if both the markers and the labels are iterables of the same dimension to zip them and plot horizontally, kind of how you can already have multiple markers?

Thanks!
Johann

**Comment by tshead2** [comment]
@johanndt - Hmmm, I haven't given it a ton of thought, but off the top of my head I think a new parameter that is stored in the legend mark and used in the HTML backend is the way to go.  I see it as comparable to the `along` parameter used by many of the other marks to control the axis used to display series.

**Comment by johanndt** [comment]
Hi @tshead2 

Upon trying a few things over the weekend I just went with manually layout out the text using `rect` as suggested in gitter:

```
canvas.legend([("Budget", line1)], rect=(0, 0, 200, 30))
canvas.legend([("Breakeven", line2)], rect=(80, 0, 200, 30))
canvas.legend([("Actual", bar)], rect=(180, 0, 200, 30))
```

Which gives me this result:
![image](https://cloud.githubusercontent.com/assets/7507499/16904078/644d7e5a-4cd2-11e6-8b8e-ccdfd9234d28.png)

Trying to lay it out automatically for now just seems a little too hard. :-) So I'm not sure if you want to keep this open then or not? I'm happy if you want to close it.

**Comment by tshead2** [comment]
It's a reasonable feature request, so I'm going to leave it open for now.

Cheers,
Tim

**Comment by tshead2** [comment]
Ideally, I'd like to use tables as the underlying mechanism for laying-out legends, which would provide several useful features like rules and gaps.  However, that leaves open the question of how to insert markers into table cells.  I'm going to explore allowing markers in text as part of our rich text functionality.

**Comment by johanndt** [comment]
For what it's worth I thought reportlab was already required! pip seems to always install it by default.

**Comment by tshead2** [comment]
Technically, it's an optional requirement that's only needed for pdf (png, mpeg) output.  I made it a requirement for pip because I figure more people would want it than not, and because pip doesn't really have a notion of optional dependencies that I'm aware of.

This change would make it a hard requirement whether generating pdf or not, about which I have reservations.

**Comment by johanndt** [comment]
Just a note - we've been successfully using the `PIL.ImageFont.getsize(legend_text)` function to work out the size of each legend and automatically do the layout and spacing. Trying to calculate the length of a piece of text seems to be a very hard problem unless you just draw it and see first.

**Comment by tshead2** [comment]
FWIW, the explicit-text-layout branch contains API to do this, and is frustratingly close to being complete.

**Comment by tshead2** [comment]
It’s way too long since the last release, moving this to the 0.15.0 release.

**Comment by tshead2** [comment]
Now that markers can be embedded in text, this is easy.  See https://github.com/sandialabs/toyplot/blob/master/design/horizontal-legends.ipynb for a couple of recipes.

Closed in e0d6510a3e4bbba081c6131709c323d859bb177a.
